### ISSUE TITLE
Fl_Terminal: initial horizontal scrollbar displayed

### ISSUE THREAD

**Issue by wcout** [fltk/fltk#930]
**Describe the bug**
Many of the programs in test folder using Fl_Terminal now show an initial horizontal scrollbar, though none would be required.

**To Reproduce**
On my machine the test program `input`, `demo`, `valuators` and `unittests` show this issue.
Here an example of `input`:
 [terminal_initial_hscrollbar](https://github.com/fltk/fltk/assets/9895976/8e994c64-260a-477a-a4ab-4f2bfa7a9d77)

**Expected behavior**
There should be no horizontal scrollbar immediately after start.

**FLTK Version**
1.4 git

**Operating System / Platform:**
Ubuntu 20.04, X11/Cairo

**Additional context**
In the screenshot above I output the values of vcols and tcols calculated in `Fl_Terminal::update_scrollbar`.
`tcols` is larger than the visible with calculated with `w_to_col(scrn_.w()`. I believe there is some discrepancy (rounding?) setting up the terminal columns, depending on the width of the widget and the way the visible columns are determined.

**Comment by erco77** [comment]
Ya, the new feature's default understandably is set to "Auto", which shows the hscrollbar if text goes beyond the widget's border, and in the case shown, it actually does go beyond.

However, since we're not used to seeing hscrollbars in these apps, I'll just set it to OFF.

**Comment by erco77** [comment]
Oh, it sounds like you're saying even when text doesn't clip off the edge, it's showing the hscroll anyway due to a possible rounding error? I guess I'll have to check...

**Comment by erco77** [comment]
Short answer: this probably solves it, without having to change the default hscroll style:
```
   // Init ringbuffer. Also creates default tabstops
   if (rows == -1 || cols == -1) {
     int newrows = h_to_row(scrn_.h());  // rows based on height
-    int newcols = w_to_col(scrn_.w());  // cols based on width
+    int newcols = w_to_col(scrn_.w())-1;// cols based on width (-1: don't overshoot - STR#930)
```
Want to dig a little deeper though, just in case.

**Comment by wcout** [comment]
Not sure if this is the real solution, though it seem to fix it.
Testing with the `input` test program, I printed out the width passed to `w_to_col()` at `Fl_Terminal::init` and then later on at `Fl_Terminal::update_scrollbar`, both taken from `scrn_.w()`:

At `init` scrn_.w() is 378, later it is 374! So that may account also for the difference: 378/8=47, 374/8=46!

**Comment by Albrecht-S** [comment]
> +    int newcols = w_to_col(scrn_.w())-1;// cols based on width (-1: don't overshoot - STR#930)

@erco77 Sorry, this is OT, but ... just as a reminder. Please don't use 'STR' for GitHub Issues (this would be misleading) and don't use '#' for STR's at all in GitHub comments because this would trigger invalid references (links) to issues or PR's.

I don't have anything to the issue itself, except: @wcout and @erco77 thanks for taking care of this. Unnecessary scrollbars, particularly in small terminal widgets, would be annoying.

**Comment by erco77** [comment]
Yeah I believe the change in `w_to_col()` values between init and "later" is due to issue #837 - the widget's ctor/init() isn't allowed to access the font system directly until draw(), so it makes a rough estimate on character width+height to get thru init (which sizes the ring buffer), then defers to getting actual font info later at draw time, when we're sure the system is actually rendering to the screen.

Has to do with applications like fluid being able to create FLTK widgets but not render them on headless systems (e.g. `fluid -c somefile.fl` inside Makefiles).

I really didn't like making that change for issue #837; no clear solution when the widget needs to know proper information about the fonts to construct its internals properly. Deferring is really messy.

Since the ctor is what creates the ring buffer, and the ring buffer's size determines the row/col count, and the app can add content to the widget before it's drawn, it starts becoming a catch-22 situation.

What I could do is resize the ring buffer /again/ in the draw() method, but by that time the application may have already added content to the terminal, so shrinking the ring buffer would cause character loss.

A lot of the issues would be solved by Fl_Terminal supporting "rewrapping" as modern terminals do. That's a goal that couldn't be achieved for 1.4.0 though, so I figured supporting xterm's behavior (and a lot of the older terminal emulators) would have to be good enough for now.

@Albrecht-S  Ha, thanks. "STR": old guys, old habits.

**Comment by wcout** [comment]
Sorry, I was not clear enough, it it just **this** change to solve the issue i.e. in the `Fl_Terminal::init_()` method move setting the box type before the `update_screen_xywh()` invocation:

````diff
diff --git a/src/Fl_Terminal.cxx b/src/Fl_Terminal.cxx
index c4eaed5e9..03e16f468 100644
--- a/src/Fl_Terminal.cxx
+++ b/src/Fl_Terminal.cxx
@@ -1264,8 +1264,9 @@ void Fl_Terminal::update_scrollbar(void) {
   unsigned int hv = hscrollbar->visible();
   // Set horiz scrollbar params
   int vcols = w_to_col(scrn_.w());    // visible cols
-  int tcols = disp_cols();            // total cols we can scroll to
+  int tcols = /*disp_cols();*/ max_line_len();         // total cols we can scroll to
   if (vcols > tcols) vcols = tcols;   // don't be larger than total
+  ::printf("vcols = %d < tcols = %d, w=%d\n", vcols, tcols, scrn_.w());
   set_scrollbar_params(hscrollbar, vcols, tcols);
   // Horiz scrollbar visibility
   if (hscrollbar_style_ == SCROLLBAR_OFF) {
@@ -3359,6 +3360,7 @@ void Fl_Terminal::init_(int X,int Y,int W,int H,const char*L,int rows,int cols,i
   oflags_         = LF_TO_CRLF;         // default: "\n" handled as "\r\n"
   // scrollbar_size must be set before scrn_
   scrollbar_size_ = 0;                  // 0 uses Fl::scrollbar_size()
+  Fl_Group::box(FL_DOWN_FRAME);
   update_screen_xywh();
   // Tabs
   tabstops_       = 0;
@@ -3398,7 +3400,6 @@ void Fl_Terminal::init_(int X,int Y,int W,int H,const char*L,int rows,int cols,i
   hscrollbar_style_ = SCROLLBAR_AUTO;
 
   resizable(0);
-  Fl_Group::box(FL_DOWN_FRAME);
   Fl_Group::color(FL_BLACK);  // black bg by default
   update_screen(true);        // update internal vars after setting screen size/font
   clear_screen_home();        // clear screen, home cursor

````

**Comment by erco77** [comment]
Ah, that's interesting, I don't think I ever would have caught that, lol..!
If that's the solution, thanks very much for that, I'll check that out.

**Comment by erco77** [comment]
I did add a few followup commits to solve some other issues with the Fl_Terminal scrollbar stuff, including adding a unittest for the scrollbar size testing (issue #931).
